src/coprocessor/dag/rpn_expr/impl_arithmetic.rs

```diff
@@ -0,0 +1,334 @@
+// Copyright 2019 TiKV Project Authors. Licensed under Apache-2.0.
+
+use cop_codegen::RpnFunction;
+use cop_datatype::FieldTypeAccessor;
+
+use super::types::RpnFnCallPayload;
+use crate::coprocessor::codec::data_type::*;
+use crate::coprocessor::codec::mysql::{Decimal, Res};
+use crate::coprocessor::codec::Error;
+use crate::coprocessor::dag::expr::EvalContext;
+use crate::coprocessor::Result;
+use cop_datatype::FieldTypeFlag;
+
+#[derive(Debug, Clone, Copy, RpnFunction)]
+#[rpn_function(args = 2)]
+pub struct RpnFnIntMod;
+
+impl RpnFnIntMod {
+    #[inline]
+    fn call(
+        _ctx: &mut EvalContext,
+        payload: RpnFnCallPayload<'_>,
+        lhs: &Option<Int>,
+        rhs: &Option<Int>,
+    ) -> Result<Option<Int>> {
+        if lhs.is_none() || rhs.is_none() {
+            return Ok(None);
+        }
+
+        let (lhs, rhs) = (lhs.unwrap(), rhs.unwrap());
+        if rhs == 0i64 {
+            return Ok(None);
+        }
+
+        let lus = payload
+            .field_type_at(0)
+            .flag()
+            .contains(FieldTypeFlag::UNSIGNED);
+
+        let rus = payload
+            .field_type_at(1)
+            .flag()
+            .contains(FieldTypeFlag::UNSIGNED);
+
+        let res = match (lus, rus) {
+            (true, true) => ((lhs as u64) % (rhs as u64)) as i64,
+            (false, false) => lhs % rhs,
+            (true, false) => ((lhs as u64) % (rhs.overflowing_abs().0 as u64)) as i64,
+            (false, true) => ((lhs.overflowing_abs().0 as u64) % (rhs as u64)) as i64,
+        };
+        Ok(Some(res))
+    }
+}
+
+#[derive(Debug, Clone, Copy, RpnFunction)]
+#[rpn_function(args = 2)]
+pub struct RpnFnRealMod;
+
+impl RpnFnRealMod {
+    #[inline]
+    fn call(
+        _ctx: &mut EvalContext,
+        _payload: RpnFnCallPayload<'_>,
+        lhs: &Option<f64>,
+        rhs: &Option<f64>,
+    ) -> Result<Option<f64>> {
+        if lhs.is_none() || rhs.is_none() {
+            return Ok(None);
+        }
+        let (lhs, rhs) = (lhs.unwrap(), rhs.unwrap());
+        if rhs == 0f64 {
+            return Ok(None);
+        }
+        Ok(Some(lhs % rhs))
+    }
+}
+
+#[derive(Debug, Clone, Copy, RpnFunction)]
+#[rpn_function(args = 2)]
+pub struct RpnFnDecimalMod;
+
+impl RpnFnDecimalMod {
+    #[inline]
+    fn call(
+        _ctx: &mut EvalContext,
+        _payload: RpnFnCallPayload<'_>,
+        lhs: &Option<Decimal>,
+        rhs: &Option<Decimal>,
+    ) -> Result<Option<Decimal>> {
+        match (lhs, rhs) {
+            (None, _) | (_, None) => Ok(None),
+            (_, Some(ref rhs)) if rhs.is_zero() => Ok(None),
+            (Some(lhs), Some(rhs)) => match lhs.to_owned() % rhs.to_owned() {
```

Comment:
Can we do this without `to_owned`? This makes performance a lot worse. Or maybe we need to modify Decimal's implementation so that `%` operation doesn't need owned values.

Reply:
OK, I will address it after https://github.com/tikv/tikv/pull/4732 merged.